pinot-core/src/main/java/com/linkedin/pinot/core/io/writer/impl/OffHeapStringStore.java

```diff
@@ -0,0 +1,235 @@
+ * copying one 'char' at a time into a CharBuffer, and keeping the offsets in the CharBuffer in another IntBuffer.
+ * Empty strings are supported.
+ *
+ * @note The class is thread-unsafe.
+ *
+ * This class has a list of OffHeapStringStore.Buffer objects. As Buffer objects get filled, new Buffer objects
+ * are added to the list. New Buffers objects have twice the capacity of the previous Buffer
+ *
+ * Within a Buffer object strings are stored as below:
+ *
+ *                  __________________________________
+ *                  |  start offset of string 1      |
+ *                  |  start offset of string 2      |
+ *                  |        .....                   |
+ *                  |  start offset of string N      |
+ *                  |                                |
+ *                  |         UNUSED                 |
+ *                  |                                |
+ *                  | STRING N .....                 |
+ *                  |          .....                 |
+ *                  |          .....                 |
+ *                  | STRING N-1                     |
+ *                  |          .....                 |
+ *                  |          .....                 |
+ *                  | STRING 0 .....                 |
+ *                  |          .....                 |
+ *                  |          .....                 |
+ *                  |________________________________|
+ *
+ *
+ * We fill the buffer as follows:
+ * - The strings are added from the bottom, each new string appearing nearer to the top of the buffer, leaving no
+ *   room between them. Each string is stored as a sequence of 'char' elements. (In Java, each char element takes 2
+ *   bytes. A string has string.length() char elements in it).
+ *
+ * - The start offsets of the strings are added from the top. Each start offset is stored as an integer, taking 4 bytes.
+ *
+ * Each time we want to add a new string, we check if we have space to add the length of the string, and the string
+ * string itself. If we do, then we compute the start offset of the new string as:
+ *
+ *    new-start-offset = (start offset of prev string added) - (length of this string)
+ *
+ * The new start offset value is stored in the offset
+ *
+ *    buffer[numStringsSoFar * 4]
+ *
+ * and the string itself is stored starting at new-start-offset
+ *
+ */
+public class OffHeapStringStore implements Closeable {
+  private static class Buffer implements Closeable {
+    private static final int INT_SIZE = V1Constants.Numbers.INTEGER_SIZE;
+    private static final int CHAR_SIZE = Character.SIZE / 8;
+
+    final PinotDataBuffer _pinotDataBuffer;
+    final ByteBuffer _byteBuffer;
+    final int _startIndex;
+    final long _size;
+
+    int _numStrings = 0;
+    int _availEndOffset;  // Exclusive
+
+    private Buffer(long size, int startIndex) {
+      if (size > Integer.MAX_VALUE) {
+        size = Integer.MAX_VALUE;
```

Comment:
Does this conflict with the preconditions check you put in allocatedirect?
 Preconditions.checkArgument(size >= 0 && size < Integer.MAX_VALUE, "bad value for size " + size);